### PR TITLE
Added support of newer additional PMP configuration and address CSRs

### PR DESCRIPTION
**Pull request by Abdulwadoodd**
This PR contains all the PMP configuration and address CSRs to support 64 PMP entries as defined by the spec

### CODE REVIEW

**Comment by Abdulwadoodd** [comment]
Hi @martinberger @jrtc27,
Kindly review the PR

File: model/riscv_csr_map.sail [hunk 1]
```diff
@@ -136,6 +136,22 @@ mapping clause csr_name_map = 0x3A0  <-> "pmpcfg0"
 mapping clause csr_name_map = 0x3A1  <-> "pmpcfg1"
 mapping clause csr_name_map = 0x3A2  <-> "pmpcfg2"
 mapping clause csr_name_map = 0x3A3  <-> "pmpcfg3"
+mapping clause csr_name_map = 0x3A0  <-> "pmpcfg0"
+mapping clause csr_name_map = 0x3A1  <-> "pmpcfg1"
+mapping clause csr_name_map = 0x3A2  <-> "pmpcfg2"
+mapping clause csr_name_map = 0x3A3  <-> "pmpcfg3"
```
Comment (jrtc27) [inline model/riscv_csr_map.sail:142]:
```suggestion
```

File: model/riscv_insts_zicsr.sail [hunk 2]
```diff
@@ -193,10 +253,23 @@ function writeCSR (csr : csreg, value : xlenbits) -> unit = {
...
+    (0x3A1, 32) => { pmpWriteCfgReg(1,  value); Some(pmpReadCfgReg(1)) },  // pmpcfg1
+    (0x3A2,  _) => { pmpWriteCfgReg(2,  value); Some(pmpReadCfgReg(2)) },  // pmpcfg2
+    (0x3A3, 32) => { pmpWriteCfgReg(3,  value); Some(pmpReadCfgReg(3)) },  // pmpcfg3
+    (0x3A4,  _) => { pmpWriteCfgReg(4,  value); Some(pmpReadCfgReg(4)) },  // pmpcfg4,
+    (0x3A5, 32) => { pmpWriteCfgReg(5,  value); Some(pmpReadCfgReg(5)) },  // pmpcfg5,
+    (0x3A6,  _) => { pmpWriteCfgReg(6,  value); Some(pmpReadCfgReg(6)) },  // pmpcfg6,
+    (0x3A7, 32) => { pmpWriteCfgReg(7,  value); Some(pmpReadCfgReg(7)) },  // pmpcfg7,
+    (0x3A8,  _) => { pmpWriteCfgReg(8,  value); Some(pmpReadCfgReg(8)) },  // pmpcfg8,
+    (0x3A9, 32) => { pmpWriteCfgReg(9,  value); Some(pmpReadCfgReg(9)) },  // pmpcfg9,
+    (0x3AA,  _) => { pmpWriteCfgReg(10, value); Some(pmpReadCfgReg(10))},  // pmpcfg10,
```
Comment (jrtc27) [inline model/riscv_insts_zicsr.sail:266]: Need a space before the }

File: model/riscv_pmp_control.sail [hunk 3]
```diff
@@ -217,11 +217,204 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
...
+      PMP_Continue =>
+  match pmpMatchEntry(addr, width, acc, priv, pmp62cfg, pmpaddr62, pmpaddr61) {
+      PMP_Success  => true,
+      PMP_Fail     => false,
+      PMP_Continue =>
+  match pmpMatchEntry(addr, width, acc, priv, pmp63cfg, pmpaddr63, pmpaddr62) {
+      PMP_Success  => true,
+      PMP_Fail     => false,
+      PMP_Continue =>
+  match priv {
```
Comment (jrtc27) [inline model/riscv_pmp_control.sail:413]: Keep the formatting of these lines?

File: model/riscv_sys_control.sail [hunk 4]
```diff
@@ -113,12 +125,61 @@ function is_CSR_defined (csr : csreg, p : Privilege) -> bool =
...
+    0x3E7 => p == Machine, // pmpaddr55
+    0x3E8 => p == Machine, // pmpaddr56
+    0x3E9 => p == Machine, // pmpaddr57
+    0x3EA => p == Machine, // pmpaddr58
+    0x3EB => p == Machine, // pmpaddr59
+    0x3EC => p == Machine, // pmpaddr60
+    0x3ED => p == Machine, // pmpaddr61
+    0x3EE => p == Machine, // pmpaddr62
+    0x3EF => p == Machine, // pmpaddr63
+
```
Comment (jrtc27) [inline model/riscv_sys_control.sail:182]:
```suggestion
```

File: model/riscv_pmp_regs.sail [hunk 5]
```diff
@@ -208,7 +382,62 @@ function pmpWriteCfgReg(n, v) = {
...
+        14 => { pmp56cfg = pmpWriteCfg(pmp56cfg, v[7 ..0]);
+                pmp57cfg = pmpWriteCfg(pmp57cfg, v[15..8]);
+                pmp58cfg = pmpWriteCfg(pmp58cfg, v[23..16]);
+                pmp59cfg = pmpWriteCfg(pmp59cfg, v[31..24]);
+                pmp60cfg = pmpWriteCfg(pmp59cfg, v[39..32]);
+                pmp61cfg = pmpWriteCfg(pmp59cfg, v[47..40]);
+                pmp62cfg = pmpWriteCfg(pmp59cfg, v[55..48]);
+                pmp63cfg = pmpWriteCfg(pmp59cfg, v[63..56]);
               }
+
```
Comment (jrtc27) [inline model/riscv_pmp_regs.sail:440]:
```suggestion
```

File: model/riscv_insts_zicsr.sail [hunk 6]
```diff
@@ -193,10 +253,23 @@ function writeCSR (csr : csreg, value : xlenbits) -> unit = {
...
+    (0x3AB, 32) => { pmpWriteCfgReg(11, value); Some(pmpReadCfgReg(11))},  // pmpcfg11,
+    (0x3AC,  _) => { pmpWriteCfgReg(12, value); Some(pmpReadCfgReg(12))},  // pmpcfg12,
+    (0x3AD, 32) => { pmpWriteCfgReg(13, value); Some(pmpReadCfgReg(13))},  // pmpcfg13,
+    (0x3AE,  _) => { pmpWriteCfgReg(14, value); Some(pmpReadCfgReg(14))},  // pmpcfg14,
+    (0x3AF, 32) => { pmpWriteCfgReg(15, value); Some(pmpReadCfgReg(15))},  // pmpcfg15,
+
```
Comment (jrtc27) [inline model/riscv_insts_zicsr.sail:272]:
```suggestion
```

File: model/riscv_csr_map.sail [hunk 7]
```diff
@@ -152,6 +168,55 @@ mapping clause csr_name_map = 0x3BC  <-> "pmpaddr12"
...
+mapping clause csr_name_map = 0x3E7  <-> "pmpaddr55"
+mapping clause csr_name_map = 0x3E8  <-> "pmpaddr56"
+mapping clause csr_name_map = 0x3E9  <-> "pmpaddr57"
+mapping clause csr_name_map = 0x3EA  <-> "pmpaddr58"
+mapping clause csr_name_map = 0x3EB  <-> "pmpaddr59"
+mapping clause csr_name_map = 0x3EC  <-> "pmpaddr60"
+mapping clause csr_name_map = 0x3ED  <-> "pmpaddr61"
+mapping clause csr_name_map = 0x3EE  <-> "pmpaddr62"
+mapping clause csr_name_map = 0x3EF  <-> "pmpaddr63"
+
```
Comment (jrtc27) [inline model/riscv_csr_map.sail:219]:
```suggestion
```

**Review by jrtc27 — COMMENTED** [review]
Haven't methodically checked for copy/paste errors but this is what I found skimming it.

I had some ideas a while back about how to reduce the code duplication, but I don't think that should block adding this support, even if it gets rather silly.

**Comment by Abdulwadoodd** [comment]
Resolved the formatting issues, empty lines and duplications

**Comment by Abdulwadoodd** [comment]
I have tested the updated SAIL model by writing tests that contain pmpcfg[4,15] and pmpaddr[16,63]. And SAIL is giving the expected result. Is there anything else to do before merging this PR?

**Comment by scottj97** [comment]
@Abdulwadoodd please squash the fixes back into the first commit, instead of adding a new commit to clean up the earlier one.

**Comment by Abdulwadoodd** [comment]
Hi @scottj97, I have squashed the changes into a single (earlier) commit.

File: model/riscv_pmp_control.sail [hunk 8]
```diff
@@ -215,13 +215,205 @@ function pmpCheck forall 'n, 'n > 0. (addr: xlenbits, width: atom('n), acc: Acce
...
+    PMP_Fail     => false,
+    PMP_Continue =>
+  match pmpMatchEntry(addr, width, acc, priv, pmp63cfg, pmpaddr63, pmpaddr62) {
     PMP_Success  => true,
     PMP_Fail     => false,
     PMP_Continue => match priv {
-                      Machine => true,
-                      _       => false
+                      Machine      => true,
+                      _            => false
```
Comment (jrtc27) [inline model/riscv_pmp_control.sail:414]: Don't reformat this

File: model/riscv_pmp_regs.sail [hunk 9]
```diff
@@ -208,7 +382,61 @@ function pmpWriteCfgReg(n, v) = {
                 pmp13cfg = pmpWriteCfg(pmp13cfg, v[47..40]);
                 pmp14cfg = pmpWriteCfg(pmp14cfg, v[55..48]);
                 pmp15cfg = pmpWriteCfg(pmp15cfg, v[63..56])
-              }
+              },
+         4 => { pmp16cfg = pmpWriteCfg(pmp16cfg, v[7 ..0]);
+                pmp17cfg = pmpWriteCfg(pmp17cfg, v[15..8]);
+                pmp18cfg = pmpWriteCfg(pmp18cfg, v[23..16]);
+                pmp19cfg = pmpWriteCfg(pmp19cfg, v[31..24]);
+                pmp20cfg = pmpWriteCfg(pmp19cfg, v[39..32]);
```
Comment (jrtc27) [inline model/riscv_pmp_regs.sail:390]: Lots of copy/paste errors in these

**Review by jrtc27 — COMMENTED** [review]
Commit message is wrong, PMP configuration and address CSRs already exist, just not the newer additional ones

File: model/riscv_pmp_regs.sail (hunk 9 above)
Comment (Abdulwadoodd) [reply model/riscv_pmp_regs.sail]: FIxed

File: model/riscv_pmp_control.sail (hunk 8 above)
Comment (Abdulwadoodd) [reply model/riscv_pmp_control.sail]: Done

**Comment by Abdulwadoodd** [comment]
> 

Commit message updated

**Comment by ptomsich** [comment]
This PR has no activity for almost 9 months.
Please help to update its status:
- @jrtc27 Have the style issues been sufficiently addressed?
- @Abdulwadoodd Please rebase onto master, so this applies without conflicts!

**Comment by jrtc27** [comment]
Style looks fine

**Comment by ptomsich** [comment]
Open questions from the golden-model meeting:
- @Abdulwadoodd Are you still maintaining this PR? Can you provide an ETA for rebasing this?

**Comment by Abdulwadoodd** [comment]
@ptomsich, thanks for the reminder. I will rebase it 2nd June.

**Comment by ptomsich** [comment]
@Abdulwadoodd Thanks for the update!  I'll keep this in the queue, so we can move it forward shortly thereafter.
@billmcspadden-riscv As discussed, please review the new version (once it comes in) for correctness relative to the specification!

**Comment by martinberger** [comment]
> I will rebase it by 2nd June.

Hello @Abdulwadoodd , any updates on this?

**Comment by Timmmm** [comment]
Since the PMP CSR addresses are nicely aligned, you can somewhat reduce the crazy amount of copy and pasting like this:

```
readCSR:
    (0b001111 @ idx : bits(6),  _) => pmpReadAddrReg(unsigned(idx)),
...
register pmpaddr  : vec(64, dec, xlenbits)
...
function pmpReadAddrReg(n : range(0, 63)) -> xlenbits = {
   ...
}
```

(`pmpReadAddrReg` is needed anyway to support PMP grain, which we have implemented an plan to upstream.)

**Comment by Timmmm** [comment]
In fact maybe hold off on this. I have implemented the approach I mentioned above which means adding support for 64 PMPs is a 1 line change instead of a gazillion line copy/paste. I'll try to send a PR soon.

**Comment by Abdulwadoodd** [comment]
It is completed by #350.

**Pull request closed without merging**